libtock/alarm_timer.c

```diff
@@ -215,33 +215,37 @@ int yield_for_with_timeout(bool* cond, uint32_t ms) {
 int gettimeasticks(struct timeval *tv, __attribute__ ((unused)) void *tzvp)
 {
   uint32_t frequency, now, seconds, remainder;
-
   alarm_internal_frequency(&frequency);
   alarm_internal_read(&now);
 
-  // The microsecond calculation will overflow in the intermediate scaling of
-  // (remainder * 1000) if the remainder is approximately greater than 4e6. Because
-  // remainder is calculated as now % frequency, we can define 0 <= remainder < frequency.
-  // This implies that the tv_usec must be of type uint64_t if frequency > 4MHz to avoid
-  // an overflow from occurring. We check this in the below assertion statement.
-  const uint32_t MAX_VALID_FREQ = 4000000;
-  assert(frequency < MAX_VALID_FREQ || sizeof(tv->tv_usec) == sizeof(uint64_t));
-
-  // Confirm frequency assumption
-  assert(frequency > 0);
+  // Confirm frequency assumptions. Since remainder is
+  // ticks % frequency, remainder <= frequency. This assertion
+  // guards against overflowing the us calculation. At worst,
+  // remainder == frequency and scaling remainder to convert
+  // to us will overflow. We confirm the greatest possible
+  // remainder value (frequency) scaled by 1e6 will fit within
+  // a uint64 and guard against division by zero.
+  assert(frequency > 0 && (frequency * 1e6) <= UINT64_MAX);
 
+  // Obtain seconds and remainder due to integer divison
   seconds   = now / frequency;
   remainder = now % frequency;
 
-  // NOTE: the drawback to this microsecond calculation is the potential loss of precision
-  // when scaling frequency / 1000 (lose 3 degrees of precision). At the time of this
-  // implementation (1/31/24), the Tock timer frequency struct provides support for
-  // frequencies such as 1KHz, 16KHz, 1MHz, etc. With such frequencies, there is not a loss
-  // of precision as the 3 least significant digits do not encode data. The only case of a lose
-  // in precision is for the frequency 32.768KHz. In this case, the loss of precision introduces ~1us
-  // of error.
-  tv->tv_sec  = seconds;
-  tv->tv_usec = (remainder * 1000) / (frequency / 1000);
+  // (ticks) * (1e6 us / s) * (s / ticks) = us
+  // Because remainder is by definition less than
+  // frequency, we must be sure to first scale remainder
+  // or else floor(remainder / frequency) = 0.
+  uint64_t us = remainder * 1e6 / frequency;
```

Comment:
I think these need to be cast to `uint64_t` to ensure they are treated as 64-bits

Reply:
The plain `1e6` here and above [probably isn't doing what you want either, and is generating a double](https://stackoverflow.com/a/40465456) (incidentally, everything being promoted to doubles might be why this passed your tests, but isn't really what you were trying to do).

I think that somewhat verbose, explicit casts are warranted given the complexities and subtleties here, and the silent fungibility of C types.